### PR TITLE
Improve Phobos style guide

### PR DESCRIPTION
**Pull request by wilzbach**
Based on the discussion at #4585, here a couple of items that are part of the "unofficial" Phobos style guide, but probably should be documented.

CC @DmitryOlshansky

### CODE REVIEW

**Comment by DmitryOlshansky** [comment]
LGTM

**Comment by ZombineDev** [comment]
LGTM too

File: dstyle.dd [hunk 1]
```diff
@@ -298,6 +299,13 @@ else if (…)
 -------------------------------
     $(LI Selective imports should have a space before and after the colon (`:`) like
     `import std.range : zip`)
+    $(LI Local, selective imports should be preferred over global imports)
+    $(LI In $(I non-templated) functions should be annotated with matching attributes (`@nogc`, `@safe`, `pure`, `nothrow`))
```
Comment (CyberShadow) [inline dstyle.dd]: In non-templated functions, WHAT should be annotated? Did you mean to not add "In"?

File: dstyle.dd [hunk 2]
```diff
@@ -298,6 +299,13 @@ else if (…)
...
+    $(LI $(I Templated functions) shouldn't be annotated with attributes as the compiler can infer them. However unittests
+         should be annotated to ensure function attributes (e.g. `pure nothrow @nogc @safe unittest { ... }`))
```
Comment (CyberShadow) [inline dstyle.dd]: Unittests should be annotated to ensure function attributes ... do what?

File: dstyle.dd [hunk 3]
```diff
@@ -298,6 +299,13 @@ else if (…)
...
+    $(LI The use of `auto` as return type should be avoided wherever possible, as it makes the documentation easier to read))
```
Comment (CyberShadow) [inline dstyle.dd]: This reads like the use of `auto` makes the documentation easier to read (then why avoid it?)

File: dstyle.dd [hunk 4]
```diff
@@ -298,6 +299,13 @@ else if (…)
...
+    $(LI `static` nested structs (aka $(LINK2 https://wiki.dlang.org/Voldemort_types, Voldemort types)) should be used for ranges)
```
Comment (CyberShadow) [inline dstyle.dd]: I would use "function-nested" and/or link to https://dlang.org/spec/struct.html#nested to clarify what the structs are nested in.

**Comment by wilzbach** [comment]
@CyberShadow I tried to address your comments. Overall I think that some of these points don't only apply for Phobos, but are part of the "D Style" in general, e.g. local selective imports or the section about attributes).

**Comment by Hackerpilot** [comment]
:white_check_mark:

**Comment by wilzbach** [comment]
Does this need approval from W/A?

(It's blocking https://github.com/dlang/phobos/pull/4585 a bit)

**Comment by wilzbach** [comment]
Ping - it's still blocking https://github.com/dlang/phobos/pull/4585 ;-)

**Comment by schuetzm** [comment]
Nobody objected, and I see nothing controversial. Merging...

**Pull request merged**